src/main/java/application/SessionController.java

```diff
@@ -166,11 +200,101 @@ public CaseFilterResponseBean filterCasesHQ(@RequestBody CaseFilterRequestBean f
     public SyncDbResponseBean syncUserDb(@RequestBody SyncDbRequestBean syncRequest) throws Exception {
+        log.info("SyncDb Request: " + syncRequest);
         syncRequest.setRestoreService(restoreService);
         String restoreXml = syncRequest.getRestoreXml();
         CaseAPIs.restoreIfNotExists(syncRequest.getUsername(), restoreXml);
         return new SyncDbResponseBean();
     }
+
+    @RequestMapping(Constants.URL_INSTALL)
+    public MenuResponseBean performInstall(@RequestBody InstallRequestBean installRequestBean) throws Exception {
+        InstallRequest installRequest = new InstallRequest(installRequestBean, restoreService, menuRepo, installService);
+        return installRequest.getResponse();
+    }
+
+    @RequestMapping(Constants.URL_MENU_SELECT)
+    public SessionBean selectMenu(@RequestBody MenuSelectBean menuSelectBean) throws Exception {
+        MenuSession menuSession = new MenuSession(menuRepo.find(menuSelectBean.getSessionId()), restoreService, installService);
+        boolean redrawing = menuSession.handleInput(menuSelectBean.getSelection());
+        menuRepo.save(menuSession.serialize());
+        Screen nextScreen;
+
+        if(!redrawing){
+            nextScreen = menuSession.getNextScreen();
+        } else{
+            nextScreen = menuSession.getCurrentScreen();
+        }
+
+        if(nextScreen instanceof MenuScreen){
+            MenuScreen menuScreen = (MenuScreen) nextScreen;
+            MenuDisplayable[] options = menuScreen.getChoices();
+            HashMap<Integer, String> optionsStrings = new HashMap<Integer, String>();
+            for(int i=0; i <options.length; i++){
+                optionsStrings.put(i, options[i].getDisplayText());
+            }
+            MenuResponseBean menuResponseBean = new MenuResponseBean();
+            menuResponseBean.setMenuType(Constants.MENU_MODULE);
+            menuResponseBean.setOptions(optionsStrings);
+            menuResponseBean.setSessionId(menuSession.getSessionId());
+            return menuResponseBean;
+        } else if (nextScreen instanceof EntityScreen){
+            EntityScreen entityScreen = (EntityScreen) nextScreen;
+
+            if (entityScreen.getCurrentScreen() instanceof EntityListSubscreen) {
+                EntityListSubscreen entityListSubscreen = (EntityListSubscreen) entityScreen.getCurrentScreen();
+                String[] rows = entityListSubscreen.getRows();
+                HashMap<Integer, String> optionsStrings = new HashMap<Integer, String>();
+                for(int i=0; i <rows.length; i++){
+                    optionsStrings.put(i, rows[i]);
+                }
+                MenuResponseBean menuResponseBean = new MenuResponseBean();
+                menuResponseBean.setMenuType(Constants.MENU_ENTITY);
+                menuResponseBean.setOptions(optionsStrings);
+                menuResponseBean.setSessionId(menuSession.getSessionId());
+                return menuResponseBean;
```

Comment:
the above 8 lines seem like they can be abstracted into something like `getMenuResponse(entityListRows, sessionId, menuType)`